changefeedccl: node crash after altering a table with an avro changefeed

**Describe the problem**

When altering a table to add a decimal type column on a table that has an Avro changefeed running on it, the node crashes after a new record with the new schema is inserted.

**To Reproduce**

```sql
CREATE EXTERNAL CONNECTION kafka1 as 'kafka://localhost:29092?topic_prefix=cdc-cockroach-'; 
CREATE EXTERNAL CONNECTION schema as 'http://schema-registry:8081';

CREATE TABLE test1(c1 int primary key, c2 int);
INSERT INTO test1 VALUES (1,1),(2,2);

CREATE CHANGEFEED FOR TABLE test1 INTO 'external://kafka1' WITH OPTIONS (avro_schema_prefix = 'crdb_cdc_', confluent_schema_registry = 'external://schema', diff, format = 'avro', on_error = 'pause', updated);
```

```sql
root@localhost:26257/defaultdb> SELECT * FROM test1;
  c1 | c2
-----+-----
   1 |  1
   2 |  2
(2 rows)
```

```
ALTER TABLE test1 ADD c3 DECIMAL(19,0);
INSERT INTO test1 VALUES (3,3,100.1);

root@localhost:26257/defaultdb> SELECT * FROM test1;
  c1 | c2 |  c3
-----+----+-------
   1 |  1 | NULL
   2 |  2 | NULL
   3 |  3 |  100
(3 rows)
```

<details>
<summary>Panic:</summary>

```
I240201 14:10:15.897872 1 util/log/flags.go:211  [-] 1  stderr capture started
panic: interface conversion: interface {} is nil, not float64
goroutine 5253 [running]:
panic({0x54ce4a0, 0xc006ce9590})
	GOROOT/src/runtime/panic.go:987 +0x3ba fp=0xc0012734b8 sp=0xc0012733f8 pc=0x49e43a
runtime.panicdottypeE(0x0, 0x523e540, 0x539c300)
	GOROOT/src/runtime/iface.go:262 +0x6a fp=0xc0012734d8 sp=0xc0012734b8 pc=0x46f86a
github.com/linkedin/goavro/v2.buildCodecForTypeDescribedByString(0x5437260?, {0xc00602e740, 0x1f}, {0xc0049963e0, 0x5}, 0x0?, 0xc001273658?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:668 +0x78d fp=0xc0012735b8 sp=0xc0012734d8 pc=0x3b2056d
github.com/linkedin/goavro/v2.buildCodecForTypeDescribedByMap(0xb90000c001273660?, {0xc00602e740, 0x1f}, 0xc006ce8f00, 0xc0055a8ae0)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:638 +0xb9 fp=0xc001273618 sp=0xc0012735b8 pc=0x3b1fc79
github.com/linkedin/goavro/v2.buildCodec(0x5437140?, {0xc00602e740?, 0x5d3d029?}, {0x5437260?, 0xc006ce8f00?}, 0x8?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:612 +0xd4 fp=0xc001273670 sp=0xc001273618 pc=0x3b1fb14
github.com/linkedin/goavro/v2.makeCodecInfo(0x20?, {0xc00602e740, 0x1f}, {0xc001ca2dc0, 0x3, 0x0?}, 0x1?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/union.go:67 +0x217 fp=0xc001273778 sp=0xc001273670 pc=0x3b30117
github.com/linkedin/goavro/v2.buildCodecForTypeDescribedBySlice(0x5437260?, {0xc00602e740, 0x1f}, {0xc001ca2dc0, 0x3, 0x4}, 0xc001273950?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/union.go:235 +0xf6 fp=0xc0012738a8 sp=0xc001273778 pc=0x3b318b6
github.com/linkedin/goavro/v2.buildCodecForTypeDescribedByMap(0x920000c00602e740?, {0xc00602e740, 0x1f}, 0xc006ce8ed0, 0xc0055a8ae0)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:642 +0x107 fp=0xc001273908 sp=0xc0012738a8 pc=0x3b1fcc7
github.com/linkedin/goavro/v2.buildCodec(0x54307e0?, {0xc00602e740?, 0xc001763880?}, {0x5437260?, 0xc006ce8ed0?}, 0x0?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:612 +0xd4 fp=0xc001273960 sp=0xc001273908 pc=0x3b1fb14
github.com/linkedin/goavro/v2.makeRecordCodec(0x54307e0?, {0xc00602e760?, 0xc0017637a0?}, 0x26?, 0x5?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/record.go:47 +0x652 fp=0xc001273ac8 sp=0xc001273960 pc=0x3b2df12
github.com/linkedin/goavro/v2.buildCodecForTypeDescribedByString(0x5437260?, {0xc00602e760, 0x1f}, {0xc004996310, 0x6}, 0x0?, 0xc001273c48?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:696 +0x52a fp=0xc001273ba8 sp=0xc001273ac8 pc=0x3b2030a
github.com/linkedin/goavro/v2.buildCodecForTypeDescribedByMap(0xf60000c001273c50?, {0xc00602e760, 0x1f}, 0xc006ce8e10, 0xc0055a8ae0)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:638 +0xb9 fp=0xc001273c08 sp=0xc001273ba8 pc=0x3b1fc79
github.com/linkedin/goavro/v2.buildCodec(0x5437140?, {0xc00602e760?, 0x5d3d029?}, {0x5437260?, 0xc006ce8e10?}, 0xc001273ce0?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:612 +0xd4 fp=0xc001273c60 sp=0xc001273c08 pc=0x3b1fb14
github.com/linkedin/goavro/v2.makeCodecInfo(0x20?, {0xc00602e760, 0x1f}, {0xc006281600, 0x2, 0x0?}, 0x1?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/union.go:67 +0x217 fp=0xc001273d68 sp=0xc001273c60 pc=0x3b30117
github.com/linkedin/goavro/v2.buildCodecForTypeDescribedBySlice(0x5437260?, {0xc00602e760, 0x1f}, {0xc006281600, 0x2, 0x2}, 0xc001273f40?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/union.go:235 +0xf6 fp=0xc001273e98 sp=0xc001273d68 pc=0x3b318b6
github.com/linkedin/goavro/v2.buildCodecForTypeDescribedByMap(0x960000c00602e760?, {0xc00602e760, 0x1f}, 0xc006ce8de0, 0xc0055a8ae0)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:642 +0x107 fp=0xc001273ef8 sp=0xc001273e98 pc=0x3b1fcc7
github.com/linkedin/goavro/v2.buildCodec(0x54307e0?, {0xc00602e760?, 0xc001763790?}, {0x5437260?, 0xc006ce8de0?}, 0x730dd78?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:612 +0xd4 fp=0xc001273f50 sp=0xc001273ef8 pc=0x3b1fb14
github.com/linkedin/goavro/v2.makeRecordCodec(0x54307e0?, {0x0?, 0xc004996150?}, 0x6?, 0xc001274120?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/record.go:47 +0x652 fp=0xc0012740b8 sp=0xc001273f50 pc=0x3b2df12
github.com/linkedin/goavro/v2.buildCodecForTypeDescribedByString(0x5437260?, {0x0, 0x0}, {0xc004996150, 0x6}, 0x472387?, 0x20?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:696 +0x52a fp=0xc001274198 sp=0xc0012740b8 pc=0x3b2030a
github.com/linkedin/goavro/v2.buildCodecForTypeDescribedByMap(0xc006ce8f60?, {0x0, 0x0}, 0xc006ce8c60, 0xc0055a8ae0)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:638 +0xb9 fp=0xc0012741f8 sp=0xc001274198 pc=0x3b1fc79
github.com/linkedin/goavro/v2.buildCodec(0xc0002b2a00?, {0x0?, 0x500?}, {0x5437260?, 0xc006ce8c60?}, 0x488b05?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:612 +0xd4 fp=0xc001274250 sp=0xc0012741f8 pc=0x3b1fb14
github.com/linkedin/goavro/v2.NewCodecFrom({0xc0002b2500, 0x4d1}, 0x4d1?)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:211 +0xf0 fp=0xc001274418 sp=0xc001274250 pc=0x3b1df30
github.com/linkedin/goavro/v2.NewCodec(...)
	github.com/linkedin/goavro/v2/external/com_github_linkedin_goavro_v2/codec.go:99
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.envelopeToAvroSchema({0xc0017633e0?, 0x0?}, {0x1, 0x1, 0x0, 0x1, 0x0}, 0xc0013cc380, 0xc0013cce70, 0x0, ...)
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/avro.go:953 +0x9ad fp=0xc0012744b8 sp=0xc001274418 pc=0x3ff4dad
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*confluentAvroEncoder).EncodeValue(0xc0059a7b00, {0x737f230, 0xc005a39890}, {{0x17afc2cabb8f87e4, 0x0, 0x0}, {0x17afc2cabb8f87e4, 0x0, 0x0}, {0x0, ...}}, ...)
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/encoder_avro.go:264 +0x5fe fp=0xc0012748d8 sp=0xc0012744b8 pc=0x4016ebe
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*kvEventToRowConsumer).encodeAndEmit(0xc00649f200, {0x737f230, 0xc005a39890}, {0xc001f395e0, {0x17afc2cabb8f87e4, 0x0, 0x0}, {0xc006d92c60, 0x6, 0x6}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/event_processing.go:446 +0x685 fp=0xc001274b70 sp=0xc0012748d8 pc=0x401f5a5
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*kvEventToRowConsumer).ConsumeEvent(0xc00649f200, {0x737f230, 0xc005a39890}, {0xc0072c99e0, 0x2, {0x0, 0x0, 0x0}, {0xc16707c7eec2b99f, 0x502d2b77ea, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/event_processing.go:387 +0xbe9 fp=0xc001274ff0 sp=0xc001274b70 pc=0x401e8e9
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changeAggregator).tick(0xc0012c8500)
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_processors.go:605 +0x21b fp=0xc0012751d8 sp=0xc001274ff0 pc=0x400179b
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changeAggregator).Next(0xc0012c8500)
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_processors.go:560 +0x6e fp=0xc001275240 sp=0xc0012751d8 pc=0x400138e
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changeFrontier).Next(0xc001ded800)
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_processors.go:1199 +0xe9 fp=0xc001275350 sp=0xc001275240 pc=0x4004969
github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run({0x737f230, 0xc005a39830}, {0x73a1ac0, 0xc001ded800}, {0x7352228?, 0xc001df3180?})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:193 +0x55 fp=0xc0012753a0 sp=0xc001275350 pc=0x227bc55
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc001ded800, {0x737f188?, 0xc005ae5580?}, {0x7352228?, 0xc001df3180})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:726 +0x73 fp=0xc0012753e8 sp=0xc0012753a0 pc=0x2280f33
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc001eca500, {0x737f188?, 0xc005ae5580}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:580 +0x238 fp=0xc0012754b8 sp=0xc0012753e8 pc=0x261f4b8
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000b1eb40, {0x737f188, 0xc005ae54c0}, 0xc001e8f4a0, 0x0, 0xc005b5f100, 0xc006177c00, 0xc001ee3000, 0xc00202ff10)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902 +0xb93 fp=0xc001275e18 sp=0xc0012754b8 pc=0x355a6b3
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.startDistChangefeed.func3({0x737f188, 0xc005ae5440})
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:328 +0x37c fp=0xc001275f58 sp=0xc001275e18 pc=0x3ffc43c
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25 fp=0xc001275f78 sp=0xc001275f58 pc=0x1b2a3a5
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64 fp=0xc001275fe0 sp=0xc001275f78 pc=0x1a80944
runtime.goexit()
	GOROOT/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc001275fe8 sp=0xc001275fe0 pc=0x4d35c1
created by golang.org/x/sync/errgroup.(*Group).Go
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
```
</details>

**Expected behavior**

The node does not crash. If this is an error, surface something appropriate to the client.

**Environment:**
 - CockroachDB version `23.1.11`

Jira issue: CRDB-35821

--- a/pkg/ccl/changefeedccl/avro.go
+++ b/pkg/ccl/changefeedccl/avro.go
@@ -75,8 +75,8 @@ const (
 type avroLogicalType struct {
 	SchemaType  avroSchemaType `json:"type"`
 	LogicalType string         `json:"logicalType"`
-	Precision   int            `json:"precision,omitempty"`
-	Scale       int            `json:"scale,omitempty"`
+	Precision   *int           `json:"precision,omitempty"`
+	Scale       *int           `json:"scale,omitempty"`
 }
 
 type avroArrayType struct {
@@ -528,8 +528,8 @@ func typeToAvroSchema(typ *types.T) (*avroSchemaField, error) {
 		decimalType := avroLogicalType{
 			SchemaType:  avroSchemaBytes,
 			LogicalType: `decimal`,
-			Precision:   prec,
-			Scale:       width,
+			Precision:   &prec,
+			Scale:       &width,
 		}
 		setNullableWithStringFallback(
 			decimalType,

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -9247,3 +9247,27 @@ func TestPubsubAttributes(t *testing.T) {
 
 	cdcTest(t, testFn, feedTestForceSink("pubsub"))
 }
+
+// TestChangefeedAvroDecimalColumnWithDiff is a regression test for
+// https://github.com/cockroachdb/cockroach/issues/118647.
+func TestChangefeedAvroDecimalColumnWithDiff(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	testFn := func(t *testing.T, s TestServer, f cdctest.TestFeedFactory) {
+		sqlDB := sqlutils.MakeSQLRunner(s.DB)
+		sqlDB.Exec(t, `CREATE TABLE test1 (c1 INT PRIMARY KEY, c2 INT, c3 DECIMAL(19, 0))`)
+		sqlDB.Exec(t, `INSERT INTO test1 VALUES (1, 2, 3);`)
+
+		schemaReg := cdctest.StartTestSchemaRegistry()
+		defer schemaReg.Close()
+		str := fmt.Sprintf(`CREATE CHANGEFEED FOR TABLE test1 WITH OPTIONS (avro_schema_prefix = 'crdb_cdc_', diff, confluent_schema_registry ="%s", format = 'avro', on_error = 'pause', updated);`, schemaReg.URL())
+		testFeed := feed(t, f, str)
+		defer closeFeed(t, testFeed)
+
+		_, ok := testFeed.(cdctest.EnterpriseTestFeed)
+		require.True(t, ok)
+	}
+
+	cdcTest(t, testFn, feedTestForceSink("kafka"))
+}

